pandas/core/frame.py

```diff
@@ -2990,15 +2990,24 @@ def duplicated(self, subset=None, keep='first'):
         from pandas.core.groupby import get_group_index
         from pandas.core.algorithms import factorize
         from pandas.hashtable import duplicated_int64, _SIZE_HINT_LIMIT
+        from pandas.compat import long
 
         def f(vals):
-
-            # if we have integers we can directly index with these
+            # if we have integers we may be able to directly index with these
             if com.is_integer_dtype(vals):
-                from pandas.core.nanops import unique1d
-                labels, shape = vals, unique1d(vals)
-            else:
-                labels, shape = factorize(vals, size_hint=min(len(self), _SIZE_HINT_LIMIT))
+                vals = vals.astype('i8', copy=False)
+
+                lower, upper = vals.min(), vals.max()
+
+                if lower >= 0:
+                    return vals, upper + 1
+                else:
+                    width = long(upper) - long(lower)
+
+                    if width < np.iinfo(np.int64).max - 1:
```

Comment:
yeh these could be in ``pandas/core/common.py`` I guess (do separate PR for this). they are already imported directly from the c-lib in cython, so this is just for python space